manta-pay/src/crypto/hash/poseidon/hasher.rs

```diff
@@ -47,18 +51,28 @@ pub trait Specification<COM = ()> {
     /// Number of Partial Rounds
     const PARTIAL_ROUNDS: usize;
 
-    /// Returns the additive identity of the field.
-    fn zero(compiler: &mut COM) -> Self::Field;
+    /// Returns the domain tag. We use different domain_tags different applications so that defending
+    /// against rainbow table attack.
+    fn domain_tag(compiler: &mut COM, arity: usize) -> Self::Field;
 
     /// Adds two field elements together.
     fn add(lhs: &Self::Field, rhs: &Self::Field, compiler: &mut COM) -> Self::Field;
 
+    /// Adds a field element with a constant
+    fn addi(lhs: &Self::Field, rhs: &Self::ParameterField, compiler: &mut COM) -> Self::Field;
```

Comment:
I have the same feeling as Tom. See https://www3.ntu.edu.sg/home/smitha/fyp_gerald/addiInstruction.html#:~:text=The%20ADDI%20instruction%20performs%20an,source%20register's%20address%2C%20immediate%20data.

Reply:
The name `addi` is short-hand for "add immediate data". We should be using full names for variables and function names and `add_const` means "add constant", which in this case is clearer as you don't have to know what "immediate data means". I would even prefer `add_constant` but I won't go that explicit since `const` is already conventional language in Rust. Also, we should not be using register metaphors for these variables, they are more algebraic. If you look at how `Self::ParameterField` is actually allocated it uses the `Constant` trait, not the "immediate data" trait, so this matches closer to its meaning and naming conventions in the constraint system language.

Even if short-hands are conventional in some scenarios, unless matching exact compatibility with those systems (like building a register machine), we should prefer to be more explicit and assume no prior knowledge rather than being implicit and assuming familiarity with systems _other than_ the ones we are building.